### PR TITLE
Move to a matrix-based build system

### PR DESCRIPTION
**Pull request by Ornias1993**
**Description of the change**

This PR is a first step in moving from seperate workflows for each container, to a single workflow for all containers.
It uses a Matrix, filled with the names of folders in `./`  that recieved changes in a push or PR and executes the build job for each of those.

It also checks if the goss.yaml file is present and executes/ignores goss testing accordingly.

**Benefits**

This significantly reduces the amount of duplicate code and makes it significantly faster/easier to deploy new containers

**Possible drawbacks**

It's ofcoarse new and requires time to mature, like every git workflow.

The only detected bug during testing, is that it does not like force-pushed squashes to master much, it doesn't break things but it tends to ignore some of the folders changed in the squash.

**Applicable issues**

**Additional information**

### CODE REVIEW

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -1,119 +1,136 @@
...
     - name: Build and Push (PR)
       if: github.event_name == 'pull_request'
       uses: docker/build-push-action@v2
       with:
         build-args: VERSION=${{ steps.prep.outputs.version }}
         context: .
-        file: ./${{ github.workflow }}/Dockerfile
-        platforms: linux/amd64,linux/arm64
+        file: ./${{ matrix.container }}/Dockerfile
+        platforms: linux/amd64
```
Comment (angelnu) [inline .github/workflows/build.yaml:98]: Missing arm64

File: .github/workflows/build.yaml [hunk 2]
```diff
@@ -1,119 +1,136 @@
...
     - name: Build and Push (non PR)
       if: github.event_name != 'pull_request'
       uses: docker/build-push-action@v2
       with:
         build-args: VERSION=${{ steps.prep.outputs.version }}
         context: .
-        file: ./${{ github.workflow }}/Dockerfile
-        platforms: linux/amd64,linux/arm64
+        file: ./${{ matrix.container }}/Dockerfile
+        platforms: linux/amd64
```
Comment (angelnu) [inline .github/workflows/build.yaml:132]: Missing arm64

**Review by angelnu — CHANGES_REQUESTED** [review]
Could you add the arm64? Maybe we should provide for an override option (like a file in the container folder) in case some containers might not work with the default target platforms.

File: .github/workflows/build.yaml (hunk 1 above)
Comment (Ornias1993) [reply .github/workflows/build.yaml:98]: Shoot, my mistake!

File: .github/workflows/build.yaml (hunk 2 above)
Comment (Ornias1993) [reply .github/workflows/build.yaml]: Fixed!

File: .github/workflows/build.yaml (hunk 1 above)
Comment (Ornias1993) [reply .github/workflows/build.yaml]: Fixed!

File: .github/workflows/build.yaml [hunk 3]
```diff
@@ -1,119 +1,136 @@
...
+    needs: [detect_changes]
+    strategy:
+      matrix: ${{fromJson(needs.detect_changes.outputs.matrix)}}
+      fail-fast: true 
     if: "!contains(github.event.head_commit.message, '[ci-skip]')"
     steps:
     - name: Checkout
       uses: actions/checkout@v2
-
+
```
Comment (onedr0p) [inline .github/workflows/build.yaml:44]: Extra space

File: .github/workflows/build.yaml [hunk 4]
```diff
@@ -1,119 +1,136 @@
...
+        else
+          echo ::set-output name=goss::false
+        fi
+ 
     - name: Set up QEMU
       uses: docker/setup-qemu-action@v1
       with:
         platforms: all
-
+
```
Comment (onedr0p) [inline .github/workflows/build.yaml:60]: Extra space

File: .github/workflows/build.yaml [hunk 5]
```diff
@@ -1,119 +1,136 @@
...
+ 
     - name: Set up Docker Buildx
       id: buildx
       uses: docker/setup-buildx-action@v1
       with:
         install: true
         version: latest
         driver-opts: image=moby/buildkit:master
-
+
```
Comment (onedr0p) [inline .github/workflows/build.yaml:68]: Extra space

File: .github/workflows/build.yaml [hunk 6]
```diff
@@ -1,119 +1,136 @@
...
+ 
     - name: Login to GitHub Container Registry
       if: github.event_name != 'pull_request'
       uses: docker/login-action@v1
       with:
         registry: ghcr.io
         username: ${{ secrets.GHCR_USERNAME }}
         password: ${{ secrets.GHCR_TOKEN }}
-
+
```
Comment (onedr0p) [inline .github/workflows/build.yaml:76]: Extra space

File: .github/workflows/build.yaml [hunk 7]
```diff
@@ -1,119 +1,136 @@
...
         password: ${{ secrets.GHCR_TOKEN }}
-
+ 
     - name: Set up goss/dgoss
       uses: e1himself/goss-installation-action@v1.0.3
+      if: ${{ steps.prep.outputs.goss == 'true' }}
       with:
         version: 'v0.3.16'
-
+
```
Comment (onedr0p) [inline .github/workflows/build.yaml:82]: Extra space

File: .github/workflows/build.yaml [hunk 8]
```diff
@@ -1,119 +1,136 @@
...
+ 
     - name: Cache Docker layers
       uses: actions/cache@v2
       with:
         path: /tmp/.buildx-cache
         key: ${{ runner.os }}-buildx-${{ github.sha }}
         restore-keys: |
           ${{ runner.os }}-buildx-
-
+
```
Comment (onedr0p) [inline .github/workflows/build.yaml:90]: Extra space

File: .github/workflows/build.yaml [hunk 9]
```diff
@@ -1,119 +1,136 @@
...
         platforms: linux/amd64,linux/arm64
+        file: ./${{ matrix.container }}/Dockerfile
         push: false # GitHub secrets not available in PRs
         tags: |
           ghcr.io/${{ github.repository_owner }}/pull-request:${{github.event.number}}
         cache-from: type=local,src=/tmp/.buildx-cache
         cache-to: type=local,dest=/tmp/.buildx-cache
-
-    - name: Build local single-platform for testing
+
```
Comment (onedr0p) [inline .github/workflows/build.yaml:104]: Extra space

File: .github/workflows/build.yaml [hunk 10]
```diff
@@ -1,119 +1,136 @@
...
+        file: ./${{ matrix.container }}/Dockerfile
         load: true
         tags: |
           ghcr.io/${{ github.repository_owner }}/pull-request:${{github.event.number}}
         cache-from: type=local,src=/tmp/.buildx-cache
         cache-to: type=local,dest=/tmp/.buildx-cache
-
-    - name: Test image
-      if: github.event.number != ''
+
```
Comment (onedr0p) [inline .github/workflows/build.yaml:117]: Extra space

File: .github/workflows/build.yaml [hunk 11]
```diff
@@ -1,119 +1,136 @@
...
         load: true
         tags: |
           ghcr.io/${{ github.repository_owner }}/pull-request:${{github.event.number}}
         cache-from: type=local,src=/tmp/.buildx-cache
         cache-to: type=local,dest=/tmp/.buildx-cache
-
-    - name: Test image
-      if: github.event.number != ''
+ 
+    - name: Test image
```
Comment (onedr0p) [inline .github/workflows/build.yaml:118]: Extra space

File: .github/workflows/build.yaml [hunk 12]
```diff
@@ -1,119 +1,136 @@
...
+ 
+    - name: Test image 
+      if: ${{ github.event.number != '' && steps.prep.outputs.goss  == 'true' }}
       env:
-        GOSS_FILE: ${{ github.workflow }}/goss.yaml
+        GOSS_FILE: ${{ matrix.container }}/goss.yaml
       run: |
         dgoss run ghcr.io/${{ github.repository_owner }}/pull-request:${{github.event.number}}
-
+
```
Comment (onedr0p) [inline .github/workflows/build.yaml:124]: Extra space

**Review by onedr0p — CHANGES_REQUESTED** [review]

**Comment by Ornias1993** [comment]
Detecting a small bug in detection of the .version file, working on a fix as we speak.

**Review by angelnu — APPROVED** [review]

**Comment by Ornias1993** [comment]
Example Run is available at TrueCharts repo:
https://github.com/truecharts/containers/actions/runs/649522119

**Comment by onedr0p** [comment]
Thanks! Let's ship it

**Pull request merged**